color is legacy

SondraE opened the pull request:
![](https://images.unsplash.com/photo-1477764250597-dffe9f601ae8?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwxfDB8MXxyYW5kb218MHx8Y3V0ZSxhbmltYWwsQ0hBTkdFX01FfHx8fHx8MTY2ODc5MDE1MA&ixlib=rb-4.0.3&q=80)
<!-- Swap `CHANGE_ME` for a random string to get a different image -->



## Description
(https://github.com/sass/sass-spec/pull/1842)


## Steps to test/reproduce
- new file called legacy
- tests to determine which colors are legacy

SondraE committed:
added tests for legacy colors and non-legacy colors

dvdherron commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,175 @@
+<===> rgb/input.scss
```
completely new functions like this won't be implemented in `libsass` so we can note that right at the beginning. 

```suggestion
<===> options.yml
---
:ignore_for:
- libsass

<===>
===================================================================
<===> rgb/input.scss
```

dvdherron commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,175 @@
+
+<===>
+================================================================================
+<===> srgb/input.scss
```
At some point we'll have to decide which of this need to be included to feel like we've covered enough cases. We probably won't need to test all of the color types.

dvdherron reviewed (commented):
This is on the right track Sondra. The format looks good. We will have to decide what tests to include (we probably won't test all color spaces?). 

I think adding some error tests would be next here (look at some of the "error/too_few_args", "error/too_many_args", "error/type" tests in the project. A lot of these follow a similar pattern.)

SondraE committed:
add note that new function won't be implemented in libsass

Co-authored-by: david herron <41588129+dvdherron@users.noreply.github.com>

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+<===> options.yml
```
@SondraE I don't know this for sure, but I am guessing the filename `is_legacy.hrx` would be preferred.

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> hwb/input.scss
+@use "sass:color";
+a {b: color.is-legacy(hwb(0, 50%, 0%))}
```
```suggestion
a {b: color.is-legacy(hwb(0 50% 0%))}
```

I believe Sass will support with or without commas

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> hsl/input.scss
+@use "sass:color";
+a {b: color.is-legacy(hsl(110, 31%, 32%))}
```
```suggestion
a {b: color.is-legacy(hsl(110 31% 32%))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> srgb/input.scss
+@use "sass:color";
+a {b: color.is-legacy(srgb(115, 20, 210))}
```
@mirisuzanne @dvdherron I've not really come across a lot with srgb used in a css notation like this before, but would it be more like:

```
a {b: color.is-legacy(color(srgb 0.45098 0.07843 0.823530))}
```

stacyk replied:
https://css.oddbird.net/sass/color-spaces/proposal/#predefined-color-spaces
> 
> ‘Predefined color spaces’ can be described using the color() function.
> The predefined RGB spaces are:
> 
> - srgb
> - srgb-linear
> - display-p3
> - a98-rgb
> - prophoto-rgb
> - rec2020

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> srgb-linear/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
@mirisuzanne @dvdherron @SondraE and same for here?

a {b: color.is-legacy(color(srgb-linear 0.45098 0.07843 0.823530))}

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+  b: false;
+}
+
+<===>
```
@SondraE I think this is the beginning of a new file so line 182 here can be removed (unless you are adding more of course).

SondraE committed:
added too many, too few, and type errors

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> display-p3/input.scss
+@use "sass:color";
+a {b: color.is-legacy((display-p3 0.515 0.35 0.3 / 1))}
```
```suggestion
a {b: color.is-legacy(color(display-p3 0.515 0.35 0.3 / 1))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> rec2020/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
```suggestion
a {b: color.is-legacy(color(rec2020 0.42053 0.979780 0.00579))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> prophoto-rgb/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
```suggestion
a {b: color.is-legacy(color(prophoto-rgb 0.42444 0.934918 0.190922))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> a98-rgb/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
```suggestion
a {b: color.is-legacy(color(a98-rgb 0 1 0))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> xyz-d50/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
```suggestion
a {b: color.is-legacy(color(xyz-d50 0.2005 0.14089 0.4472))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> xyz-d65/input.scss
+@use "sass:color";
+a {b: color.is-legacy(?????)}
```
```suggestion
a {b: color.is-legacy(color(xyz-d65 0.21661 0.14602 0.59452))}
```

stacyk commented on spec/core_functions/color/legacy.hrx:
```diff
@@ -0,0 +1,182 @@
+================================================================================
+<===> oklab/input.scss
+@use "sass:color";
+a {b: color.is-legacy(oklab(0.44027, 0.08818, -0.13386))}
```
```suggestion
a {b: color.is-legacy(oklab(0.44027 0.08818 -0.13386))}
```

I believe the commas would be an error for oklab, oklch,lab, and lch

stacyk reviewed (commented):
@SondraE looks great so far. I added some suggestions as I believe commas are no longer needed in the color notations. 

I don't know how many color formats we will want to test but it may make sense to include something like:

- a named color
- hex color
- maybe something with alpha in the legacy color space?

SondraE committed:
revised based on Stacy's comments about commas, file name, and adding hex, name, and alpha colors

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+  |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^ invocation
+  '
+  ,--> sass:color
+1 | @function color.is-legacy(rgb($amount $amount $amount)) {
```
I think instead of testing how the `is-legacy()` function works this is instead testing if the correct values are in the `rgb()` function. For `is-legacy()`, the "too few" error should be triggered with an empty function (`color.is-legacy()`) since this function only takes one argument (`$color`).

And then for all of the errors in this file, see Natalie's [note](https://github.com/sass/sass-spec/pull/1841#discussion_r1028545449) about leaving the error message content blank.

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+<===>
+================================================================================
+<===> error/too_many_args/input.scss
+a {b: color.is-legacy(rgb(0 255 0 15))}
```
Same here. We're just trying to test `is-legacy()`. A "too many" args error should be triggered if we run this function with more than one color:

`a {b: color.is-legacy(rgb(0 255 0), blue)}`

(see the functions for blue() and blackness() for examples of errors for a single argument function)

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+  input.scss 1:7  root stylesheet
+
+<===> error/too_many_args/error-libsass
+Error: wrong number of arguments (4 for 3) for `color.is-legacy'
```
Since this probably won't ever be implemented in libsass, we can omit any libsass error messages for this function and any entirely new function we write tests for. At the top of the file we have told the test runner to ignore libsass completely with `:ignore_for: - libsass`

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+<===>
+================================================================================
+<===> error/type/input.scss
+a {b: color.is-legacy(rgb(0 255 red))}
```
```suggestion
a {b: color.is-legacy(1)}
```
This function expects a color. A number would be the wrong type for the $color argument. 

It might be good to also write an additional `type` test (`error/type/string/input.scss`) to make sure this errors if $color is a string (in quotes):

`a {b: color.is-legacy("blue")}`

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+
+<===>
+================================================================================
+<===> lch/input.scss
```
```suggestion
<===> named/input.scss
```
We could swap out one of the tests or add an additonal test called `named`. This will test that the function also works if the argument is explicitly named e.g. `$color: lch(90.6 52.8 197)`

dvdherron commented on spec/core_functions/color/is_legacy.hrx:
```diff
@@ -0,0 +1,284 @@
+================================================================================
+<===> lch/input.scss
+@use "sass:color";
+a {b: color.is-legacy(lch(90.6 52.8 197))}
```
```suggestion
a {b: color.is-legacy($color: lch(90.6 52.8 197))}
```

SondraE committed:
removed unneeded error message info and fixed named and rec

stacyk reviewed (approved):
@SondraE These updates look good to me. I think this is ready to move forward....which I think means opening a PR on the sass/sass-spec repo, right @dvdherron? We are go over that tomorrow if so if you want.

dvdherron commented:
> @SondraE These updates look good to me. I think this is ready to move forward....which I think means opening a PR on the sass/sass-spec repo, right @dvdherron? We are go over that tomorrow if so if you want.

Missed this but it looks like PR was opened successfully? :) 🚀

SondraE committed:
added missing error inputs

SondraE committed:
switched blue to a on too many arguments error and added use sass color to errors

nex3 committed:
[Color 4] Update tests for Color Level 4 (#1846)

This updates most tests for the currently-implemented chunk of the
Color 4 feature. A few tests that need more thorough reworking (such
as being moved out of `error/` for things that are no longer intended
to be errors) are marked as "todo" instead.

See #1828
See sass/sass#2831

dvdherron committed:
Merge branch 'feature.color-4' into color-is-legacy

nex3 committed:
Generate error outputs

nex3 committed:
Poke CI

stacyk committed:
fix too few and named tests

nex3 committed:
Use feature branches for testing the JS API (#1853)

This also skips a few color tests for the embedded host

nex3 committed:
Merge branch 'feature.color-4' of github.com:sass/sass-spec into color-is-legacy

nex3 committed:
Update output for too_few_args
